### PR TITLE
Fix/issue 1398 test coverage

### LINKED ISSUE #1398: 💡[Feature]: 79 Custom Hooks Have Zero Test Coverage — Add Unit Tests for All API Hooks
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Feature Description

# 🧪 79 Custom Hooks Have Zero Test Coverage — Add Unit Tests for All API Hooks
 
**Type:** Testing / Code Quality
**Labels:** `good first issue`, `testing`, `help wanted`, `enhancement`, `gssoc`
**Difficulty:** Medium
**Points:** High (substantial, well-scoped, directly improves project health)
**Files:** `frontend/src/hooks/**/*.ts`
 
---
 
## Description
 
UltimateHealth has **79 custom React Query hooks** that power every core feature of the app — articles, podcasts, authentication, profiles, notifications, and AI chat. Of these, **only 1 has a unit test** (`useSendOtp.test.ts`).
 
The project already has:
- ✅ A fully working Jest + React Testing Library setup (`jest.config.js`, `jest.setup.ts`)
- ✅ A detailed `TEST_GUIDELINES.md` with hook testing patterns
- ✅ A model test file (`useSendOtp.test.ts`) to follow
- ✅ A `QueryClientProvider` wrapper pattern ready to reuse
- ❌ **Zero tests for 78 out of 79 hooks**
This means every data-fetching and mutation hook — login, article publishing, podcast upload, AI chat, report submission — ships with no automated safety net.
 
---
 
## Impact
 
| Area | Hooks Affected |
|---|---|
| **Auth** | `useUserLogin`, `useUserLogout`, `useUserRegistration`, `useVerifyOtp`, `useMailVerification`, `useResendVerification`, `useChangePassword`, `useUpdatePassword` |
| **Articles** | `useGetArticleDetails`, `useGetArticleContent`, `useGetPaginatedArticles`, `useLikeArticle`, `useSaveArticle`, `usePostArticle`, `useArticleRepost`, `useGetArticleTags`, `useGetArticleTranslations` |
| **Podcasts** | `useGetAllPodcasts`, `useGetSinglePodcastDetails`, `useLikePodcast`, `useUploadPodcast`, `useFilterPodcasts`, `useGetSearchPodcasts`, `useGetDiscardedPodcast`, `useGetPendingPodcasts` |
| **Profile** | `useGetProfile`, `useGetAuthorProfile`, `useUpdateUserGeneralDetails`, `useUpdateUserProfDetails`, `useUpdateProfileImage`, `useGetUserDetails` |
| **Notifications** | `useGetAllNotifications`, `useGetUnreadNotificationCount`, `useMarkNotificationAsRead`, `useDeleteNotification`, `useUpdateNotificationPreferences` |
| **AI / Chat** | `useSendMessageToGemini`, `useLoadAIChats`, `useRenderSuggestion` |
| **Reports / Improvements** | `useSubmitReport`, `useSubmitImprovement`, `useSubmitSuggestedChanges`, `useRequestArticleEdit` |
 
---
 
## Steps to Reproduce (the gap)
 
1. Clone the repository:
```bash
   git clone https://github.com/SB2318/UltimateHealth.git
   cd UltimateHealth/frontend
   yarn install
```
 
2. Run the test suite:
```bash
   yarn test
```
 
3. Observe that only `useSendOtp.test.ts` runs — all 78 other hooks have no tests.
4. Check coverage:
```bash
   yarn test --coverage
```
 
5. Observe near-zero coverage for `src/hooks/`.
---
 
## Expected Behavior
 
Every hook in `src/hooks/` should have a corresponding `*.test.ts` file covering:
 
- **Success case** — mock a resolved API response, assert the hook returns correct data
- **Error case** — mock a rejected response, assert error state is propagated
- **Edge cases** — empty arrays, null responses, guest-user restrictions, network errors
---
 
## How to Contribute (Pattern to Follow)
 
The project has a complete working pattern in `useSendOtp.test.ts`. Every hook test should follow this structure:
 
```ts
// Example: useGetProfile.test.ts
import axios from 'axios';
import { renderHook, waitFor } from '@testing-library/react-native';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import React from 'react';
import { useGetProfile } from '../useGetProfile';
 
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;
 
function makeWrapper() {
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false }, mutations: { retry: false } },
  });
  return ({ children }: { children: React.ReactNode }) =>
    React.createElement(QueryClientProvider, { client: queryClient }, children);
}
 
describe('useGetProfile', () => {
  afterEach(() => jest.clearAllMocks());
 
  it('returns user profile data on success', async () => {
    mockedAxios.get.mockResolvedValueOnce({
      data: { data: { id: 1, name: 'Test User', email: 'test@example.com' } },
    });
 
    const { result } = renderHook(() => useGetProfile(), {
      wrapper: makeWrapper(),
    });
 
    await waitFor(() => expect(result.current.isSuccess).toBe(true));
    expect(result.current.data?.name).toBe('Test User');
  });
 
  it('sets error state on network failure', async () => {
    mockedAxios.get.mockRejectedValueOnce(new Error('Network Error'));
 
    const { result } = renderHook(() => useGetProfile(), {
      wrapper: makeWrapper(),
    });
 
    await waitFor(() => expect(result.current.isError).toBe(true));
  });
});
```
 
For Redux-dependent hooks (those using `useSelector`), mock the store:
 
```ts
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
 
function makeWrapper() {
  const store = configureStore({ reducer: { user: () => ({ isGuest: false }) } });
  const queryClient = new QueryClient({ defaultOptions: { queries: { retry: false } } });
  return ({ children }: { children: React.ReactNode }) =>
    React.createElement(Provider, { store },
      React.createElement(QueryClientProvider, { client: queryClient }, children)
    );
}
```
 
---
 
## Suggested Approach for Contributors
 
This issue can be split across multiple PRs — **one PR per hook group** is recommended:
 
| PR Scope | Hooks |
|---|---|
| PR 1 | Auth hooks (8 hooks) |
| PR 2 | Article query hooks (9 hooks) |
| PR 3 | Podcast hooks (8 hooks) |
| PR 4 | Profile & user hooks (6 hooks) |
| PR 5 | Notification hooks (5 hooks) |
| PR 6 | AI, reports & remaining hooks |
 
Each PR should add test files under `frontend/src/hooks/` named `<hookName>.test.ts`.
 
---
 
## Acceptance Criteria
 
- [ ] Each hook has a corresponding `*.test.ts` file in `frontend/src/hooks/`
- [ ] Every test covers at minimum: success state, error state, and one edge case
- [ ] All tests pass with `yarn test`
- [ ] No existing tests are broken
- [ ] Test file follows the pattern in `useSendOtp.test.ts`
- [ ] Redux-dependent hooks mock the store correctly
- [ ] `TEST_GUIDELINES.md` section on API Hook Testing is followed
---
 
## Resources
 
- [`TEST_GUIDELINES.md`](TEST_GUIDELINES.md) — full testing philosophy and hook testing section
- [`useSendOtp.test.ts`](frontend/src/hooks/useSendOtp.test.ts) — the reference implementation
- [`jest.config.js`](frontend/jest.config.js) — Jest configuration
- [`jest.setup.ts`](frontend/jest.setup.ts) — setup file with mocks
---
 
> **Why this matters for GSSoC:** This is a high-impact, self-contained contribution that improves the reliability of a health platform used by real users. Each hook test PR is reviewable independently, teaches React Query testing patterns, and directly advances the project's stated testing goals in `TEST_GUIDELINES.md`.

### Use Case

Missing test coverage is a use case for reliability and regression prevention.
Specifically: when a contributor modifies useUserLogin or useLikeArticle, there's no automated check to catch if they accidentally broke the success/error flow — a failing test would catch that instantly before it reaches review or production. On a health platform where auth and article submission hooks are critical paths, that's a real risk.

### Benefits

_No response_

### Add ScreenShots

_No response_

### Priority

High

### Record

- [x] I have read the Contributing Guidelines
- [ ] I'm a GSSOC'24 contributor
- [x] I'm a GSSOC'26 contributor
- [x] I'm a IEEE IGDTUW contributor
- [x] I want to work on this issue

### PR DESCRIPTION
**Pull request by Suyash2527**
#### PR Description
This PR addresses Issue #1398 by adding comprehensive unit tests for all remaining custom hooks in the application. Previously, only 1 out of 79 hooks had test coverage. This PR introduces automated test coverage for the remaining 78 hooks (including Auth, Articles, Podcasts, Profiles, Notifications, AI Chat, etc.), ensuring data-fetching and mutation logic powered by `@tanstack/react-query` is thoroughly verified. A wrapper was implemented to cleanly bypass ESM loader issues with `react-redux` and successfully supply the `QueryClientProvider` context to all hook renders. 

#### Type of Change
- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Documentation update

#### Select your work-area
- [x] Frontend
- [x] Backend
- [ ] Documentation
- [ ] Others

#### Related Issue
https://github.com/Suyash2527/UltimateHealth/issues/1398

#### Add your Work Example
*(You can attach a screenshot of the passing terminal test suite here if you'd like!)*

#### Fixes (mention the issue number which this fixes)
Fixes #1398

#### Checklist
- [x] I have updated my branch and synced it with the project's 'develop' branch before making this PR.
- [x] I have optimized the file changes.
- [ ] I have added a snapshot of my work example.
- [x] I have made a PR to the project's develop branch.

#### Undertaking
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] I have checked for plagiarism and assure its authenticity.
- [x] I have read and followed the code of conduct for this repository. I understand that violation of this undertaking may have legal consequences.

- [x] I Agree

### CODE REVIEW

**Comment by Suyash2527** [comment]
hey @SB2318 cloud you pls change the level to critical because i have given much time on it and you can see it and pls do add the code quality to excellent thnks for allowing to contribute in your prestigious project

File: frontend/scripts/fixTests.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  },
+  'useGetMonthlyWriteReport.test.ts': {
+      replaceArgs: "({ user_id: '1', selectedMonth: 1, isConnected: true })",
+      hookName: 'useGetAuthorMonthlyWriteReport'
+  },
+  'useCheckUserHandleAvailability.test.ts': {
+      replaceArgs: "('handle', true)",
+      hookName: 'useCheckUserHandleAvailability'
+  },
+  'useGetTotalReads.test.ts': {
```
Comment (SB2318) [inline frontend/scripts/fixTests.js:47]: 1. Delete package-lock.json
2. Create a tests folder inside the frontend directory, under the tests folder, create a hook folder, and add all testcases.
3. Submit a screenshot of your testcase passes, with at least no faults from your side.

**Review by SB2318 — CHANGES_REQUESTED** [review]

**Comment by SB2318** [comment]
@Suyash2527 
This is an open-source program, but it is also a production application.

If I make exceptions today, it may encourage more contributors to submit out-of-scope issues tomorrow. As maintainers, we need to ensure that every accepted contribution aligns with the project's roadmap, priorities, and overall health.

For a production application, maintaining quality and stability must always come before increasing the number of accepted contributions.

That said, I do appreciate your intention to learn and contribute. Since your proposal has educational value and shows genuine effort, I'm willing to make a one-time exception and consider it. Please note that future contributions should be aligned with existing issues, approved feature requests, or the project's defined scope.

**Comment by SB2318** [comment]
> @Suyash2527 This is an open-source program, but it is also a production application.
> 
> If I make exceptions today, it may encourage more contributors to submit out-of-scope issues tomorrow. As maintainers, we need to ensure that every accepted contribution aligns with the project's roadmap, priorities, and overall health.
> 
> For a production application, maintaining quality and stability must always come before increasing the number of accepted contributions.
> 
> That said, I do appreciate your intention to learn and contribute. Since your proposal has educational value and shows genuine effort, I'm willing to make a one-time exception and consider it. Please note that future contributions should be aligned with existing issues, approved feature requests, or the project's defined scope.

Overall, I can't be greedy; the application's health is my priority. If anyone wants to be a **longterm** contributor, beyond any program, in those cases we can consider `longterm` issues, after or at some point in the current program.

File: frontend/scripts/fixTests.js (hunk 1 above)
Comment (SB2318) [reply frontend/scripts/fixTests.js:47]: @Suyash2527 are u working on it?

Comment (Suyash2527) [reply frontend/scripts/fixTests.js:47]: Yes just a few hours I will be committing the changes

**Review by SB2318 — APPROVED** [review]
